### PR TITLE
Browser-extension popup interface + app modes

### LINKED ISSUE #94: Annotation display switch in browser popup.

### LINKED ISSUE #95: "add annotation" mode

### PR DESCRIPTION
**Pull request by kaglowka**
resolves #94 
resolves #95

I'm sorry, this PR has grown much beyond the size comfortable to review. A clear distinction could have been done e.g. between the interface and the business logic and let's consider it a recommendation for the future.

**summary**
- new browser popup interface
- features that can be turned on in the popup:
    - "add annotation" mode
    - "turn off the extension"
    - "turn off the extension for this page"
- an annotation mode widget visible in the annotation mode
- chrome storage used for communication between browser popup and content scripts
- temporarily, for "client" configuration, chrome storage is mocked for this configuration not to stop working altogether; this is a quick temporary solution, that we should work on or drop soon. For now, annotation mode is always enabled in client configuration.
- From [two available kind of Chrome storages](https://developer.chrome.com/apps/storage), `local` was chosen for development and `sync` for production due to [quota on sync usage per browser extension](https://developer.chrome.com/extensions/storage#type-StorageArea)

**Several changes still need to be done or considered**
#124
#126
#127

### CODE REVIEW

File: src/index.tsx [hunk 1]
```diff
@@ -34,9 +39,44 @@ function loadInitialData() {
   store.dispatch(readEndpoint('/annotations?url=' + window.location.href));
 }
 
+export function loadDataFromChromeStorage() {
```
Comment (glowka) [inline src/index.tsx]: All other actions taking place here are just one-liners and this function take most of the file, is it that important? How about moving it to less prominent place?

File: src/store/appModes/reducers.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+const initialState: AppModeReducer = {
+  isExtensionDisabled: false,
+  annotationModePages: [],
+  disabledPages: [],
+};
+
+export default function appModes(state = initialState, action) {
+  switch (action.type) {
+    case MODIFY_APP_MODES:
+      return appModesActionHandler(state, action.payload);
```
Comment (glowka) [inline src/store/appModes/reducers.ts]: Calling this trivial function does not make it more descriptive to me and it definitely gets harder to read with every action being trivial function that I have to follow just to find it's trivial. Do we need this layer? Why not using `{ ...state,  ...payload,  }` directly?

File: src/store/reducer.ts [hunk 3]
```diff
@@ -1,14 +1,16 @@
 import { combineReducers } from 'redux';
 import { reducer as api } from 'redux-json-api';
 import widgets, { WidgetReducer } from './widgets/reducers';
+import appModes, { AppModeReducer } from './appModes/reducers';
 import textSelector from './textSelector/reducers';
 import { AnnotationAPIModel, AnnotationUpvoteAPIModel } from 'api/annotations';
 
-export interface IStore {
+export interface IAppState {
```
Comment (glowka) [inline src/store/reducer.ts]: Nice change but maybe go even further and be even more descriptive like `IPageState` or `ITabState`?

File: src/browser-extension/popup/BrowserPopup.tsx [hunk 4]
```diff
@@ -1,41 +1,180 @@
...
+
+    this.loadStateFromStorage();
+  }
+
+  componentDidUpdate() {
+    this.loadStateFromStorage();
+  }
+
+  isExtensionDisabledForCurrentTab() {
+    return (this.state.disabledPages || []).indexOf(this.state.currentTabUrl) !== -1;
```
Comment (lgodziejewski) [inline src/browser-extension/popup/BrowserPopup.tsx]: I think using [includes](https://devdocs.io/javascript/global_objects/array/includes) (as it is finally available) is a bit more readable

File: src/browser-extension/popup/BrowserPopup.tsx [hunk 5]
```diff
@@ -1,41 +1,180 @@
...
+      currentTabUrl,
+    } = this.state;
+
+    const isAnnotationMode = this.isAnnotationModeForCurrentTab();
+    if (!isAnnotationMode) {
+      let newAnnotationModePages;
+      newAnnotationModePages = [...annotationModePages, currentTabUrl];
+      this.setState({ annotationModePages: newAnnotationModePages });
+      chromeStorage.set({ [chromeKeys.ANNOTATION_MODE_PAGES]: newAnnotationModePages });
+      window.close();
```
Comment (lgodziejewski) [inline src/browser-extension/popup/BrowserPopup.tsx]: doesn't this close the tab?

File: src/browser-extension/popup/BrowserPopup.tsx [hunk 6]
```diff
@@ -1,41 +1,180 @@
...
+    }
+  }
+
+  handleDisabledExtensionChange = (e: ChangeEvent<HTMLInputElement>) => {
+    const newValue = e.target.checked;
+    this.setState({ isExtensionDisabled: newValue });
+    chromeStorage.set({ [chromeKeys.DISABLED_EXTENSION]: newValue });
```
Comment (lgodziejewski) [inline src/browser-extension/popup/BrowserPopup.tsx]: as the `setState` and `chromeStorage.set` pair is repeated here and there, maybe it'd be good to wrap them in a function (e.g. `setPersistedState`)?

File: src/browser-extension/popup/toggle/toggle.tsx [hunk 7]
```diff
@@ -0,0 +1,30 @@
+import React from 'react';
+
+import './toggle.scss';
+
+declare const chrome: any;
```
Comment (lgodziejewski) [inline src/browser-extension/popup/toggle/toggle.tsx]: is it needed here?

File: src/components/editor/Editor.tsx [hunk 8]
```diff
@@ -294,7 +299,7 @@ class Editor extends React.Component<
           className={styles.close}
           onClick={this.onCancelClick}
         >
-          <i className="remove icon" />
+          <i className="remove icon"/>
```
Comment (lgodziejewski) [inline src/components/editor/Editor.tsx]: Maybe we should configure tslint to define a style for stuff like this (and below)? (unless it already is configured and I'm not aware of it 😛)

File: src/index.tsx [hunk 9]
```diff
@@ -34,9 +39,44 @@ function loadInitialData() {
...
+  return new Promise((resolve, reject) => {
+    chromeStorage.get([
+      chromeKeys.ANNOTATION_MODE_PAGES,
+      chromeKeys.DISABLED_EXTENSION,
+      chromeKeys.DISABLED_PAGES,
+    ], (result) => {
+      console.log(result);
```
Comment (lgodziejewski) [inline src/index.tsx]: .

File: src/index.tsx (hunk 1 above)
Comment (lgodziejewski) [reply src/index.tsx]: 👍, I'd say it should be moved to some utils/actions/ file

File: src/init/chromeStorageHandlers.ts [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+
+const storageKeysToAppMode = {
+  [chromeKeys.DISABLED_EXTENSION]: 'isExtensionDisabled',
+  [chromeKeys.ANNOTATION_MODE_PAGES]: 'annotationModePages',
+  [chromeKeys.DISABLED_PAGES]: 'disabledPages',
+};
+
+export default function initializeChromeStorageHandlers() {
+  chromeStorage.onChanged.addListener((changes, namespace) => {
+    const newModes = {};
```
Comment (lgodziejewski) [inline src/init/chromeStorageHandlers.ts]: unused?

File: src/init/chromeStorageHandlers.ts [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+    for (const key of Object.keys(changes)) {
+      store.dispatch(
+        changeAppModes({ [storageKeysToAppMode[key]]: changes[key].newValue } as AppModeReducer),
```
Comment (lgodziejewski) [inline src/init/chromeStorageHandlers.ts]: maybe it'd be better to construct `newModes` as
```
newModes = Object.keys(changes).reduce(acc, key => acc[key] = changes[key].newValue, {})
```
and then use a single `store.dispatch` call with the whole object?

Now there is a separate dispatch for every `key` available in `changes`

File: src/store/appModes/reducers.ts (hunk 2 above)
Comment (lgodziejewski) [reply src/store/appModes/reducers.ts]: Usually action handlers are split into functions if they are a bit more complex (and if there's a lot of them), but in this case I'd say we can do it like @glowka suggested, especially that it's rather very simple.

Unless we want to have a strict style of writing reducers (maybe better for general code readability - same solutions used everywhere, one decision less to make when writing new code)

File: src/store/appModes/selectors.ts [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+  state => state.appModes,
+  (appModes) => {
+    // Standardize the URL by disregarding stuff that does not identify a page like URL parameters etc.
+    const currentStandardizedUrl = standardizeUrlForPageSettings(window.location.href);
+
+    return {
+    // Page highlights are disabled when
+    //  the extension is disabled in general or when it is disabled for this particular page
+      arePageHighlightsDisabled: appModes.isExtensionDisabled
+      || appModes.disabledPages.indexOf(currentStandardizedUrl) !== -1,
```
Comment (lgodziejewski) [inline src/store/appModes/selectors.ts]: `includes`

File: src/utils/url.js [hunk 13]
```diff
@@ -30,3 +29,8 @@ export function httpPrefixed(url) {
 export function isValidUrl(url) {
   return url.match(urlRegex)
 }
+
+export function standardizeUrlForPageSettings(url) {
```
Comment (lgodziejewski) [inline src/utils/url.js]: There is a nice npm package for url operations like this - https://www.npmjs.com/package/url-parse (and some of the functions above)

**Comment by kaglowka** [comment]
> however please keep it shorter in the future

Couldn't agree more :)

File: src/browser-extension/popup/BrowserPopup.tsx [hunk 14]
```diff
@@ -1,41 +1,180 @@
...
+          isExtensionDisabled: result[chromeKeys.DISABLED_EXTENSION] || false,
+          disabledPages: result[chromeKeys.DISABLED_PAGES] || [],
+        });
+      });
+    }
+
+  }
+
+  componentDidMount() {
+    chrome.tabs.query({ active: true, currentWindow: true }, (tabs) => {
```
[the remark this answers is not in the document]
Comment (kaglowka) [reply src/browser-extension/popup/BrowserPopup.tsx]: I completely agree -- I didn't want to come into the browser-independent overhead just yet.

File: src/browser-extension/popup/BrowserPopup.tsx (hunk 4 above)
Comment (kaglowka) [reply src/browser-extension/popup/BrowserPopup.tsx]: Thanks, didn't know about it :+1:

File: src/browser-extension/popup/BrowserPopup.tsx (hunk 5 above)
Comment (kaglowka) [reply src/browser-extension/popup/BrowserPopup.tsx]: BrowserPopup lives in a completely separate environment, and `window` is in this case the browser popup area.

File: src/init/chromeStorageHandlers.ts (hunk 10 above)
Comment (kaglowka) [reply src/init/chromeStorageHandlers.ts]: :+1:

File: src/utils/url.js (hunk 13 above)
Comment (kaglowka) [reply src/utils/url.js]: Thanks, it's good to know.

**Pull request merged**